### PR TITLE
Parser fixes

### LINKED ISSUE #301: Potentially unnecessary deprecation notice with v3.0.0rc2
## 🐛 Bug Report

Using the following `tartufo.toml` config file, I would not expect to see any deprecation notices.

```toml
[tool.tartufo]
exclude-path-patterns = [
	{ path-pattern = 'tartufo.toml', reason = '...' },
	{ path-pattern = 'package-lock.json', reason = '...' },
	{ path-pattern = 'server/package-lock.json', reason = '...' },
	{ path-pattern = 'client/package-lock.json', reason = '...' },
]
```

However, running the following command `tartufo scan-local-repo .`, produces the following output, am I potentially missing something here?

```sh
/usr/local/lib/python3.9/site-packages/tartufo/scanner.py:325: DeprecationWarning: Old format of --exclude-path-patterns option and config file setup exclude-path-patterns = ['exclusion pattern'] has been deprecated and will be removed in a future version. Make sure all the exclusions are set up using new pattern i.e. exclude-path-patterns = [{path-pattern='exclusion pattern',reason='reason for exclusion'}] in the config file
  warnings.warn(
Time: 2021-12-10T15:36:53.587169
All clear. No secrets detected.
```

## Expected Behavior

I would not expect to see deprecation notices.

## Environment

```
tartufo, version 3.0.0rc2
macOS 12.0.1 (21A559)
Python 3.9.9
```

### PR DESCRIPTION
**Pull request by rbailey-godaddy**
To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [ ] Tests (if applicable)
* [ ] Documentation (if applicable)
* [X] Changelog entry
* [X] A full explanation here in the PR description of the work done

## PR Type
What kind of change does this PR introduce?

* [X] Bugfix
* [ ] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the API in any way?

* [X] Yes (backward compatible)
* [ ] No (breaking changes)

## Issue Linking

closes #301 

## What's new?
- Declare compound options as `type=click.UNPROCESSED`. Otherwise the default type is `click.STRING` and new-style global option values are returned as strings and confused with old-style values. This did not affect target repository options, which are parsed directly using `tomlkit`.
- Fix comparison error that was supposed to avoid duplicate processing of options when the target repository is the current repository (but didn't because the types were different).
- Convert tartufo's `pyproject.toml` so we don't generate deprecation warnings about ourselves.
- As a side-effect of the first change, and previous work, I verified that old-style and new-style values can be intermixed freely within the same configuration file (including within the same option) without problems. I don't recommend we promote or even advertise this behavior.

### CODE REVIEW

**Review by sushantmimani — APPROVED** [review]
LGTM 🦅

File: pyproject.toml [hunk 1]
```diff
@@ -73,81 +73,81 @@ docs = ["recommonmark", "sphinx", "sphinx-click", "sphinx-rtd-theme", "sphinxcon
...
-    'tests/data/testRules.json',
-    'tartufo/data/default_regexes\.json',
-    '\.git/.*',
+    {path-pattern = '\.github/workflows/(.*)\.yml'},
+    {path-pattern = 'docs/source/(.*)\.rst'},
+    {path-pattern = 'tests/test_base_scanner\.py'},
+    {path-pattern = 'tests/test_util.py'},
+    {path-pattern = 'tests/data/testRules.json'},
+    {path-pattern = 'tartufo/data/default_regexes\.json'},
+    {path-pattern = '\.git/.*'},
```
Comment (tarkatronic) [inline pyproject.toml]: Since we're in here touching these, can we go ahead and make sure all the `.`s are escaped?

File: pyproject.toml [hunk 2]
```diff
@@ -73,81 +73,81 @@ docs = ["recommonmark", "sphinx", "sphinx-click", "sphinx-rtd-theme", "sphinxcon
...
-  "9f675d4d7f43e484e07fb199a1e7ec372ef204631c1f4ec054acc3119fbde4cf", # build_and_publish_docker_file 13-Jun-2019
-  "88152f1d10077f417ee5aea5fa86adaa5b74d3f2de2e87ce4b4aa5cd79c7b2f5", # nothing file, 30-Dec-2016
+  {signature = "6c70b4a712ada69f2a6c919edc783cf485b6637e3c5429de376ef1a451a042e4", reason = "tests/data/scan_folder/donotscan.txt"},
+  {signature = "357841766ca69979608e4e7fb1da2addeafe6d15cade15466c39d903cc0488a8", reason = "tests/data/scan_folder/test.txt"},
+  {signature = "743b893e30777d9c4e28d3d341ca4ba1c2541a9c62b497dece75c2a64420e770", reason = "tests/test_folder_scanner.py"},
+  {signature = "358a1ae040ab71b3aafeb4318e5d9e318ff1b681c5bf0e99d381584a33ee7833", reason = "tests/data/scan_folder/scan_sub_folder/sub_folder_test.txt"},
+  {signature = "9d1b278fa384c9dc58d607130d97740910309d13b29f13f7ed6348738ea4f32c", reason = "tests/test_folder_scanner.py 16-Nov-2021"},
+  {signature = "9f675d4d7f43e484e07fb199a1e7ec372ef204631c1f4ec054acc3119fbde4cf", reason = "build_and_publish_docker_file 13-Jun-2019"},
+  {signature = "88152f1d10077f417ee5aea5fa86adaa5b74d3f2de2e87ce4b4aa5cd79c7b2f5", reason = "nothing file, 30-Dec-2016"},
```
Comment (tarkatronic) [inline pyproject.toml:150]: We should review these at some point for what's still relevant, given the path pattern and entropy exclusions... but, you know, problem for future us.

**Review by tarkatronic — APPROVED** [review]
Thanks for fixing this  up @rbailey-godaddy!

File: pyproject.toml (hunk 1 above)
Comment (rbailey-godaddy) [reply pyproject.toml]: Done!

**Pull request merged**